flocker/node/agents/test/test_blockdevice.py

```diff
@@ -995,6 +1043,222 @@ class UnusableAPI(object):
     """
 
 
+def make_icalculater_tests(calculater_factory):
+    """
+    Make a test case to test an ``ICalculater`` implementation.
+
+    :param calculater_factory: Factory to make an ``ICalculater`` provider.
+    :type calculater_factory: No argument ``callable``.
+
+    :return: A ``TestCase`` subclass.
+    """
+    class ICalculaterTests(SynchronousTestCase):
+        """
+        Tests of an ``ICalculater`` implementation.
+        """
+        def test_interface(self):
+            """
+            The ``ICalculater`` implemention actually implements the interface.
+            """
+            verifyObject(ICalculater, calculater_factory())
+
+        @given(
+            discovered_datasets=builds(
+                dataset_map_from_iterable,
+                lists(DISCOVERED_DATASET_STRATEGY),
+            ),
+            desired_datasets=builds(
+                dataset_map_from_iterable,
+                lists(DESIRED_DATASET_STRATEGY),
+            ),
+        )
+        def test_returns_changes(self, discovered_datasets, desired_datasets):
+            """
+            ``ICalculater.calculate_changes_for_datasets`` returns a
+            ``IStateChange``.
+            """
+            calculater = calculater_factory()
+            changes = calculater.calculate_changes_for_datasets(
+                discovered_datasets=discovered_datasets,
+                desired_datasets=desired_datasets)
+            self.assertTrue(IStateChange.providedBy(changes))
```

Comment:
No.
1. There is going to be a fake implementation whose return value doesn't depend on the arguments.
2. It may take several changes to result in the desired dataset. See `test_simple_transitions` below to see some initial tests that the transitions eventually result in the desired state.